ledger/from-archive/src/main.rs

```diff
@@ -41,54 +51,242 @@ fn main() {
+            return;
+        }
         if let Some(block_limit) = config.num_blocks {
-            if block_index >= block_limit {
-                log::debug!(
+            if start >= block_limit {
+                log::info!(
                     logger,
                     "Done fetching transactions for {} blocks",
-                    block_index,
+                    block_limit,
                 );
                 return;
             }
         }
-        // Construct URL for the block we are trying to fetch.
-        let filename = block_num_to_s3block_path(block_index)
-            .into_os_string()
-            .into_string()
-            .unwrap();
-        let url = transactions_fetcher.source_urls[0].join(&filename).unwrap();
-
-        // Try and get the block.
-        log::debug!(
-            logger,
-            "Attempting to fetch block {} from {}",
-            block_index,
-            url
+        log::info!(logger, "Getting block contents for {}-{}", start, end);
+        let block_map = get_block_contents(
+            transactions_fetcher.clone(),
+            start.clone(),
+            end,
+            std::time::Duration::from_secs(1),
+            &logger,
         );
-        match transactions_fetcher.block_from_url(&url) {
-            Ok(block_data) => {
-                // Append new data to the ledger
+        log::info!(logger, "Got block contents. Now appending to ledger");
+        for (block_index, block_data_opt) in block_map.iter() {
+            if let Some(block_data) = block_data_opt {
                 local_ledger
                     .append_block(
-                        block_data.block(),
-                        block_data.contents(),
+                        &block_data.block(),
+                        &block_data.contents(),
                         block_data.signature().clone(),
                     )
-                    .unwrap_or_else(|_| panic!("Could not append block {:?}", block_index))
-            }
-            Err(err) => {
-                log::debug!(
-                    logger,
-                    "Done fetching transactions for {} blocks ({:?})",
-                    block_index,
-                    err
-                );
-                return;
+                    .unwrap();
+            } else {
+                log::warn!(logger, "No block data for index {}", block_index);
+                reached_end = true;
             }
         }
-        block_index += 1;
+        start = end;
+    }
+}
+
+/// Gets all transactions for a range of block indices.
+///
+/// # Arguments
+/// * `transactions_fetcher` - The mechanism used for fetching transaction
+///   contents for a given
+/// block.
+/// * `block_start` - Start index
+/// * `block_end` - End index
+/// * `timeout` - Overall request timeout.
+///
+/// Peers are queried concurrently. Currently, this method will run indefinitely
+/// until all transactions have been retrieved.
+fn get_block_contents(
```

Comment:
Any reason to copy this entire thing instead of making the one in ledger_sync_service `pub`?

Reply:
Hm, I see that it is less strict so using the one in ledger_sync_service would be difficult.